### ISSUE TITLE
Segmentation fault with archive due to fat LTO + other optimizations

### ISSUE THREAD

**Issue by chennin** [crunchy-labs/crunchy-cli#329]
**Describe the bug**
Segmentation fault in `memchr::memmem::Finder::into_owned` when running `crunchy-cli archive`.

Hello, I have a strange crash that prevents archiving episodes that has to do with a certain combination of optimizations and architectures. I ran in to this because I installed from AUR and my RUSTFLAGS are set by default to enable build optimizations.

Although I have traced the crash to `memchr`, I am starting with this bug report in `crunchy-cli` because (1) it's likely a problem for many crunchy-cli users who build it themselves with certain optimizations, and (2) crunchy-cli is the only way I know how to replicate the problem, as I don't have a smaller reproduction. I do however have a core and backtrace. 

The main question for crunchy-cli is if you want to reduce the default LTO level (see workarounds below); and the question to anyone is if you want to help narrow the problem and get a smaller repro that could perhaps be reported to  memchr instead.

I have replicated the problem on Arch Linux (updated as of 2024-02-04) and Ubuntu 23.10 on two **znver3** systems, AND on a **skylake** system, but *not* on a znver4 system.

Log: 
[ccli-debug-log.txt](https://github.com/crunchy-labs/crunchy-cli/files/14171705/ccli-debug-log.txt)

The core (and built binary) are too big to upload here, even zipped (built with debug info, crunchy-cli is 154M), but I can find another place to upload if you want them.

<details><summary>The backtrace</summary>

```c
(gdb) bt
0  0x00005ab927678034 in memchr::memmem::Finder::into_owned (self=...)
    at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/memchr-2.7.1/src/memmem/mod.rs:474
#1  regex_automata::util::prefilter::memmem::Memmem::new<&[u8]> (needles=Python Exception <class 'gdb.error'>: value has been optimized out
&[&[u8]](size=1), _kind=<optimized out>)
    at src/util/prefilter/memmem.rs:29
#2  regex_automata::util::prefilter::Choice::new<&[u8]> (needles=Python Exception <class 'gdb.error'>: value has been optimized out
&[&[u8]](size=1), kind=<optimized out>)
    at src/util/prefilter/mod.rs:588
#3  regex_automata::util::prefilter::Prefilter::new<&[u8]> (kind=<optimized out>, needles=Python Exception <class 'gdb.error'>: value has been optimized out
&[&[u8]](size=1))
    at src/util/prefilter/mod.rs:205
#4  0x00005ab92763e2d3 in regex_automata::meta::strategy::ReverseSuffix::new (core=...,
    hirs=&[&regex_syntax::hir::Hir](size=1) = {...}) at src/meta/strategy.rs:1188
#5  regex_automata::meta::strategy::new (info=<optimized out>, hirs=&[&regex_syntax::hir::Hir](size=1) = {...})
    at src/meta/strategy.rs:170
#6  0x00005ab9276592a2 in regex_automata::meta::regex::Builder::build_many_from_hir<regex_syntax::hir::Hir> (
    self=0x7ffd3d3d7eb8, hirs=...) at src/meta/regex.rs:3553
#7  regex_automata::meta::regex::Builder::build_many<&str> (self=0x7ffd3d3d7eb8, patterns=&[&str](size=1) = {...})
    at src/meta/regex.rs:3441
#8  regex_automata::meta::regex::Builder::build (self=0x7ffd3d3d7eb8, pattern=...) at src/meta/regex.rs:3360
#9  0x00005ab927635600 in regex::builders::Builder::build_one_string (self=0x7ffd3d3d7f80) at src/builders.rs:79
#10 regex::builders::string::RegexBuilder::build (self=0x7ffd3d3d7f80) at src/builders.rs:233
#11 regex::regex::string::Regex::new (re=...) at src/regex/string.rs:181
#12 0x00005ab92749d158 in crunchy_cli_core::utils::download::get_video_stats (path=...)
    at crunchy-cli-core/src/utils/download.rs:915
#13 0x00005ab92726f39f in crunchy_cli_core::utils::download::{impl#2}::download::{async_fn#0} ()
    at crunchy-cli-core/src/utils/download.rs:241
#14 0x00005ab92722bd60 in crunchy_cli_core::archive::command::{impl#0}::execute::{async_fn#0} ()
    at crunchy-cli-core/src/archive/command.rs:262
#15 0x00005ab927220b18 in crunchy_cli_core::execute_executor::{async_fn#0}<crunchy_cli_core::archive::command::Archive>
    () at crunchy-cli-core/src/lib.rs:220
#16 crunchy_cli_core::cli_entrypoint::{async_fn#0} () at crunchy-cli-core/src/lib.rs:205
--Type <RET> for more, q to quit, c to continue without paging--
#17 0x00005ab927422b80 in crunchy_cli::main::{async_block#0} () at src/main.rs:11
#18 tokio::runtime::park::{impl#4}::block_on::{closure#0}<crunchy_cli::main::{async_block_env#0}> ()
    at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/park.rs:282
#19 tokio::runtime::coop::with_budget<core::task::poll::Poll<()>, tokio::runtime::park::{impl#4}::block_on::{closure_env#0}<crunchy_cli::main::{async_block_env#0}>> (budget=..., f=...)
    at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/coop.rs:107
#20 tokio::runtime::coop::budget<core::task::poll::Poll<()>, tokio::runtime::park::{impl#4}::block_on::{closure_env#0}<crunchy_cli::main::{async_block_env#0}>> (f=...)
    at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/coop.rs:73
#21 tokio::runtime::park::CachedParkThread::block_on<crunchy_cli::main::{async_block_env#0}> (
    f=<error reading variable: Cannot access memory at address 0x2e90>, self=<optimized out>)
    at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/park.rs:282
#22 tokio::runtime::context::blocking::BlockingRegionGuard::block_on<crunchy_cli::main::{async_block_env#0}> (
    self=<optimized out>, f=<error reading variable: Cannot access memory at address 0x2e90>)
    at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/context/blocking.rs:66
#23 tokio::runtime::scheduler::multi_thread::{impl#0}::block_on::{closure#0}<crunchy_cli::main::{async_block_env#0}> (
    blocking=0x7ffd3d3e7a28)
    at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/scheduler/multi_thread/mod.rs:87
#24 tokio::runtime::context::runtime::enter_runtime<tokio::runtime::scheduler::multi_thread::{impl#0}::block_on::{closure_env#0}<crunchy_cli::main::{async_block_env#0}>, ()> (handle=0x7ffd3d3e7a70, allow_block_in_place=true, f=...)
    at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/context/runtime.rs:65
#25 tokio::runtime::scheduler::multi_thread::MultiThread::block_on<crunchy_cli::main::{async_block_env#0}> (
    self=<optimized out>, handle=<optimized out>,
    future=<error reading variable: Cannot access memory at address 0x2e90>)
    at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/scheduler/multi_thread/mod.rs:86
#26 tokio::runtime::runtime::Runtime::block_on<crunchy_cli::main::{async_block_env#0}> (self=0x7ffd3d3e7a40,
    future=...) at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.1/src/runtime/runtime.rs:350
#27 crunchy_cli::main () at src/main.rs:11
```
</details>

**To Reproduce**

On an affected CPU, set up your crunchy-cli saved login/token, then run:

```sh
sudo sysctl -w kernel.core_pattern=core-%e-%s-%u-%g-%p-%t # Optional depending on your setup
export RUSTFLAGS="-C opt-level=2 -C target-cpu=native"
export CARGO_PROFILE_RELEASE_LTO=true # "true" is the default for crunchy-cli
export CARGO_PROFILE_RELEASE_DEBUG=true 
export CARGO_PROFILE_RELEASE_STRIP=false # Get useful info out of the core
cargo build --release
./target/release/crunchy-cli archive "https://www.crunchyroll.com/watch/GR7908EZ6/pv-1-shiro"
```

Here is a full docker/podman repro to start from scratch - again seems to be CPU-dependent, though:

<details><summary>Docker repro</summary>

The following mounts your `$HOME/.config/crunchy-cli` read-only into the container, or you could set up crunchy-cli auth within the container.

```sh
sudo sysctl -w kernel.core_pattern=core-%e-%s-%u-%g-%p-%t # Optional depending on your setup - good idea in docker to avoid systemd's core interception
podman run --rm -it -v $HOME/.config/crunchy-cli:/root/.config/crunchy-cli:ro docker.io/archlinux:base-20240101.0.204074
```

In the container:
```sh
cd /tmp && pacman-key --init >/dev/null 2>&1 && \
pacman -Syuq --noconfirm git gcc rustup gdb pkgconf ffmpeg pipewire pipewire-jack && \
rustup default 1.75.0 && \
export DEBUGINFOD_URLS="https://debuginfod.archlinux.org/" && \
git clone --depth 1 --branch v3.2.5 https://github.com/crunchy-labs/crunchy-cli.git && cd crunchy-cli
```

Then to build & test:
```sh
export RUSTFLAGS="-C opt-level=2 -C target-cpu=native"
export CARGO_PROFILE_RELEASE_LTO=true #  "true" is the default for crunchy-cli
export CARGO_PROFILE_RELEASE_DEBUG=true 
export CARGO_PROFILE_RELEASE_STRIP=false # Get useful info out of the core
cargo build --release
./target/release/crunchy-cli archive "https://www.crunchyroll.com/watch/GR7908EZ6/pv-1-shiro"
```
</details>

**To workaround**

*Any* one of the following works around the issue:

* `export CARGO_PROFILE_RELEASE_LTO=thin`, or false, or off - this *could* be where crunchy-cli could avoid the problem; ie change the setting in Cargo.toml instead of the environment variable
* Rustflag `-C opt-level=1`, or z; instead of 2 or 3
* Rustflag `-C target-cpu=x86-64` instead of native

So it seems to be some confluence of low-level optimizations, but I for one am definitely not qualified to get that deep into it.

**Versions**

* crunchy-cli 3.2.5
* Rust 1.75.0

**Comment by chennin** [comment]
Argh - I spent all that time isolating which flags caused the crash, and didn't think to try a different Rust version. **The problem goes away with rustc 1.76.0-beta.9 and 1.77.0-beta.1**, at least. (crunchy-cli does not compile with 1.74.1). 

So, nevermind?

**Comment by bytedream** [comment]
Yea I think the issue can be closed, Rust 1.76 will be released in two days. But nevertheless thanks for your error report :)
